eth_abi/registry.py

```diff
@@ -0,0 +1,127 @@
+import functools
+import operator
+
+from .exceptions import (
+    NoEntryFound,
+)
+
+
+class PredicateMapping:
+    """
+    Acts as a mapping from predicate functions to values.  Values are retrieved when their
+    corresponding predicate matches a set of given inputs.  Predicates can also be labeled to
+    facilitate removal from the mapping.
+    """
+    def __init__(self):
+        self._values = {}
+        self._labeled_predicates = {}
+
+    def add(self, predicate, value, label=None):
+        if predicate in self._values:
+            raise ValueError('Predicate already exists')
+
+        if label is not None:
+            if label in self._labeled_predicates:
+                raise ValueError('Predicate with label "{}" already exists'.format(label))
+
+            self._labeled_predicates[label] = predicate
+
+        self._values[predicate] = value
+
+    def find(self, type_str):
+        results = tuple(
+            value for predicate, value in self._values.items()
+            if predicate(type_str)
+        )
+
+        if len(results) == 0:
+            raise ValueError('No matching values found')
+
+        if len(results) > 1:
+            raise ValueError('Multiple matching values found')
+
+        return results[0]
+
+    def remove_by_reference(self, predicate):
+        # Delete the predicate mapping to the previously stored value
+        try:
+            del self._values[predicate]
+        except KeyError:
+            raise KeyError('Predicate not found')
```

Comment:
Seems like including the `str` representation of the `predicate` in the error here would be beneficial under the hope that our predicate functions are at least sometimes named in helpful ways.

Reply:
When you say string representation, do you just mean `str(func_or_lambda)`, so something like this: `<function foo at 0x110b416a8>` or `<function <lambda> at 0x110b38bf8>`?

Reply:
Just want to make sure you're not talking about using something like the `inspect` module. :)

Reply:
Just ignore the comments I made in reply to this.  I hadn't read your comment thoroughly enough.

Reply:
You are correct that I mean `str(func_or_lambda)`.  I think this would be well complimented by changing the `equals` function as follows.

```python
def equals(value):
    def inner(other):
        return other == value
    inner.__name__ = "is_equal_to_{0}".format(value)
    return inner
```

This way our auto-generated functions will have helpful string representations.

One thing that isn't ideal about this is that `equals('arst') != equals('arst')` since the generated functions won't be equivalent.  I think we can fix this with some class shenanigans.

```python
class Equals:
    def __init__(self, value):
        self.value = value
    def __call__(self, other):
        return self.value == other
    def __eq__(self, other):
        return isinstance(other, Equals) and self.value == other.value
```

There are probably a few more dunder methods that need to be implemented like `__hash__`.

Reply:
It is possible that we can forego all of that by supplying labels for the default predicates?  My thinking was that, if people wanted to unregister default behavior, they would do so via one of the informatively chosen labels that we include.  I was also thinking of having a method on the registry that will provide a nice summary of what predicates are registered and any labels they might have.

If not, I think the idea of a callable `Equals` class is kinda cool anyway and I'd be happy to go with it.  It just seems like we could potentially avoid some of the edge cases that might come out of that approach by deferring to the labeling functionality.

Thoughts?

Reply:
>  I was also thinking of having a method on the registry that will provide a nice summary of what predicates are registered and any labels they might have.

Implementing this with `__str__` seems appropriate.

Reply:
Went ahead and just took your suggestion here of having the `Equals` class.